File: CoreJava/algorithms/src/main/java/org/gb/sample/algo/intro_to_algo_book/matrix_multiplication/Matrix.java
package org.gb.sample.algo.intro_to_algo_book.matrix_multiplication;

class Matrix {

	private final String parenthesizedName;
	private final String simpleProductName;
	private final int rowCount;
	private final int columnCount;
	private final int cumulativeMultiplyCount;

	Matrix(String name, int rowCount, int columnCount) {
		this.parenthesizedName = name;
		this.simpleProductName = name;
		this.rowCount = rowCount;
		this.columnCount = columnCount;
		this.cumulativeMultiplyCount = 0;
	}

	Matrix(Matrix leftPredecessor, Matrix rightPredecessor, int rowCount, int columnCount, int cumulativeMultiplyCount) {
		StringBuilder parenthesizedNameBuilder = new StringBuilder("(");
		this.parenthesizedName = parenthesizedNameBuilder.append(leftPredecessor.getParenthesizedName()).append(".")
				.append(rightPredecessor.getParenthesizedName()).append(")").toString();
		StringBuilder simpleProductNameBuilder = new StringBuilder(leftPredecessor.getSimpleProductName());
		this.simpleProductName = simpleProductNameBuilder.append(rightPredecessor.getSimpleProductName()).toString();
		this.rowCount = rowCount;
		this.columnCount = columnCount;
		this.cumulativeMultiplyCount = cumulativeMultiplyCount;
	}

	String getParenthesizedName() {
		return parenthesizedName;
	}

	String getSimpleProductName() {
		return simpleProductName;
	}

	int getRowCount() {
		return rowCount;
	}

	int getColumnCount() {
		return columnCount;
	}

	int getCumulativeMultiplyCount() {
		return cumulativeMultiplyCount;
	}

	@Override
	public String toString() {
		String summaryInfo = "Name: %s, dimensions: %d x %d.";
		return String.format(summaryInfo, simpleProductName, rowCount, columnCount);
	}
}

File: CoreScala/simple-scala/src/main/scala/org/gb/sample/scala/intro_to_algo_book/matrix_multiplication/Matrix.scala
package org.gb.sample.scala.intro_to_algo_book.matrix_multiplication

class Matrix private(val parenthesizedName: String, val simpleProductName: String, val rowCount: Int, val columnCount: Int, val cumulativeMultiplyCount: Int) {
    override def toString: String = {
        val summaryInfo = "Name: %s, dimensions: %d x %d."
        String.format(summaryInfo, simpleProductName, rowCount, columnCount)
    }
}

object Matrix {
    def apply(name: String, rowCount: Int, columnCount: Int): Matrix = {
        new Matrix(name, name, rowCount, columnCount, 0)
    }

    def apply(leftPredecessor: Matrix, rightPredecessor: Matrix, rowCount: Int, columnCount: Int, cumulativeMultiplyCount: Int): Matrix = {
        val parenthesizedNameBuilder = new StringBuilder("(")
        val simpleProductNameBuilder = new StringBuilder(leftPredecessor.simpleProductName)
        val parenthesizedName = parenthesizedNameBuilder.append(leftPredecessor.parenthesizedName)
          .append(".")
          .append(rightPredecessor.parenthesizedName).append(")").toString()
        val simpleProductName = simpleProductNameBuilder.append(rightPredecessor.simpleProductName).toString
        new Matrix(parenthesizedName, simpleProductName, rowCount, columnCount, cumulativeMultiplyCount)
    }
}
